File: src/main/java/pl/mmakos/advent/year2024/Day3.java
package pl.mmakos.advent.year2024;

import lombok.AccessLevel;
import lombok.NoArgsConstructor;
import pl.mmakos.advent.utils.Bool;

import java.util.regex.Pattern;

import static java.lang.Integer.parseInt;
import static pl.mmakos.advent.utils.Utils.read;
import static pl.mmakos.advent.utils.Utils.stream;

@NoArgsConstructor(access = AccessLevel.NONE)
public final class Day3 {
  private static final String MUL_REGEX = "mul\\((\\d{1,3}),(\\d{1,3})\\)";

  @SuppressWarnings("java:S106")
  public static void main(String[] args) {
    System.out.println("Advent of code 2024, day 2");
    System.out.println("TASK 1: " + task1());
    System.out.println("TASK 2: " + task2());
  }

  private static int solve(String input, boolean task2) {
    String regex = MUL_REGEX;
    if (task2) regex += "|do\\(\\)|don't\\(\\)";
    Pattern pattern = Pattern.compile(regex);
    Bool enabled = new Bool(true);
    return stream(pattern.matcher(input))
        .filter(m -> {
          if (m.group().startsWith("don")) {
            enabled.set(false);
          } else if (m.group().startsWith("do")) enabled.set(true);
          return enabled.get() && !m.group().startsWith("do");
        })
        .mapToInt(m -> parseInt(m.group(1)) * parseInt(m.group(2)))
        .sum();
  }

  private static long task1() {
    return solve(read(), false);
  }

  private static long task2() {
    return solve(read(), true);
  }
}

File: src/main/cpp/year2024/day3.cpp
#include <iostream>
#include <vector>

#include <utils.h>

constexpr int day = 3;
constexpr int year = 2024;

const std::string mul = R"(mul\((\d{1,3}),(\d{1,3})\))";

long long solve(const aoc::input &input, bool task2=false) {
  std::string inputStr = aoc::join(input, "\n");
  std::regex regex( task2 ? mul + R"(|do\(\)|don't\(\))" : mul);

  auto endIt = std::sregex_iterator();
  long long sum = 0;
  bool enabled = true;
  for (auto i = std::sregex_iterator(inputStr.begin(), inputStr.end(), regex); i != endIt; ++i) {
    const std::smatch &match = *i;
    if (match[0] == "don't()") {
      enabled = false;
    } else if (match[0] == "do()") {
      enabled = true;
    } else if (enabled) {
      sum += stoi(match[1]) * stoi(match[2]);
    }
  }

  return sum;
}

long long solve1(const aoc::input &input) {
  return solve(input);
}

long long solve2(const aoc::input &input) {
  return solve(input, true);
}

int main() {
  const aoc::input input = aoc::readInput<std::string>(year, day);

  std::cout << "Advent of code " << year << ", day " << day << std::endl;
  std::cout << "TASK 1: " << solve1(input) << std::endl;
  std::cout << "TASK 2: " << solve2(input) << std::endl;
}
